### PR TITLE
[MPS] Fix linear backward reshape->matmul->reshape > 4D issue

### LINKED ISSUE #187201: MPS: linear backward crashes on 5-D input (macOS 27)
Initially discovered when trying to run MPS tests so by default it crashes the whole run (test_output_grad_match_nn_functional_linear_mps_float16 / _float32). Had experienced similar fault on macOS 26 too although it doesn't seem to show in torch CI so maybe not the same one although to be kept in mind.

Minimal repro:
```code
import torch
x = torch.randn(2, 1, 2, 1, 2, device="mps", requires_grad=True)
w = torch.randn(4, 2, device="mps", requires_grad=True)
torch.nn.functional.linear(x, w).sum().backward()
```

Crash:
```text
MPSGraphExecutable.mm:6125: failed assertion `Error: MLIR pass manager failed'
Hard abort (SIGABRT) on the metal gpu stream thread, inside mps_linear_backward -> MPSGraphExecutable applyOptimizationPassesWithDevice -> Metal __assert_rtn.
```

See attached full logs:

[MPS_SEGFAULT.md](https://github.com/user-attachments/files/28893419/MPS_SEGFAULT.md)

Tested on latest HEAD (c02cc46)

cc @ezyang @gchanan @kadeng @msaroufim @albanD @gqchen @nikitaved @soulitzer @Varal7 @bobrenjc93 @kulinseth @malfet @DenisVieriu97 @jhavukainen @aditvenk @Isalia20

### PR DESCRIPTION
**Pull request by jhavukainen**
Fixes: https://github.com/pytorch/pytorch/issues/187201

The issue stems from MPSGraph optimizing pass over the reshape->matmul->reshape which seems to misbehave when there are more than one singleton dimension in the original shape. 

EDIT: It was actually enough to just gate the macos27 to use matmul directly without reshaping. The newer OS can handle the > 4D matmul directly and avoiding the reshape+matmul+reshape pattern prevents the error on the beta 1 seed of macos27.

Once https://github.com/pytorch/pytorch/pull/186541 lands I think we'll regardless move the linear ops to use the metal kernels so we'll skip this issue

### CODE REVIEW

File: aten/src/ATen/native/mps/operations/Linear.mm [hunk 1]
```diff
@@ -234,34 +234,30 @@ static Tensor _mps_linear_backward_input(IntArrayRef input_size, const Tensor& g
     return output;
   }
 
+  // For >4D inputs, flatten to 2D outside the graph. The 2D flattening is required to avoid an older
+  // matmul bug (https://github.com/pytorch/pytorch/issues/114942), but doing it inside the graph
+  // (reshape -> matmul -> reshape) crashes the MLIR pass manager on macOS 26+
+  // (https://github.com/pytorch/pytorch/issues/187201). <=4D keeps the direct N-D matmul as before.
+  const bool needs_reshape = grad_output.dim() > 4;
```
Comment (Skylion007) [inline aten/src/ATen/native/mps/operations/Linear.mm]: Is this still bug on the latest MPS or can we version gate it?

Comment (jhavukainen) [reply aten/src/ATen/native/mps/operations/Linear.mm]: It looks like on MacOS27 at least the matmul is already working for the > 4D cases. So you're right we can just gate the reshape for pre-macos27. This retains the previous behavior for earlier releases so no perf regression concerns. 

Committed a change that'll add the MacOS27 to the OS helpers and adds the gating to linear backward.

**Comment by jhavukainen** [comment]
Ok seems like the failing flow is just the test infra unable to prepare the containers to run the tests. Unlikely to be related to the change.

```
Pod mt-l-arm64g4-16-62-2fbdp-runner-25wmv-workflow: 
phase=Pending, conditions=[PodReadyToStartContainers=True, Initialized=True, Ready=False, ContainersReady=False, PodScheduled=True], waiting=[job: ImagePullBackOff] (7208s/7200s)

Error: 
Error: pod failed to come online with error: Error: Pod mt-l-arm64g4-16-62-2fbdp-runner-25wmv-workflow is unhealthy with phase status Pending: backoff timeout
```

**Review by malfet — APPROVED** [review]
